--- array.c
+++ array.c
@@ -958 +958 @@
-static VALUE
+VALUE
@@ -1191 +1191 @@
-static VALUE
+VALUE
@@ -1234 +1234 @@
-static VALUE
+VALUE
@@ -1307 +1307 @@
-static VALUE
+VALUE
@@ -1433 +1433 @@
-static VALUE
+VALUE
@@ -1604 +1604 @@
-static VALUE
+VALUE
@@ -1668 +1668 @@
-static VALUE
+VALUE
@@ -1722 +1722 @@
-static VALUE
+VALUE
@@ -1774 +1774 @@
-static VALUE
+VALUE
@@ -1980 +1980 @@
-static VALUE
+VALUE
@@ -2029 +2029 @@
-static VALUE
+VALUE
@@ -2053 +2053 @@
-static VALUE
+VALUE
@@ -2110 +2110 @@
-static VALUE
+VALUE
@@ -2137 +2137 @@
-static VALUE
+VALUE
@@ -2165 +2165 @@
-static VALUE
+VALUE
@@ -2181 +2181 @@
-static VALUE
+VALUE
@@ -2356 +2356 @@
-static VALUE
+VALUE
@@ -2402 +2402 @@
-static VALUE
+VALUE
@@ -2424 +2424 @@
-static VALUE
+VALUE
@@ -2453 +2453 @@
-static VALUE
+VALUE
@@ -2484 +2484 @@
-static VALUE
+VALUE
@@ -2527 +2527 @@
-static VALUE
+VALUE
@@ -2543 +2543 @@
-static VALUE
+VALUE
@@ -2605 +2605 @@
-static VALUE
+VALUE
@@ -2630 +2630 @@
-static VALUE
+VALUE
@@ -2825 +2825 @@
-static VALUE rb_ary_bsearch_index(VALUE ary);
+VALUE rb_ary_bsearch_index(VALUE ary);
@@ -2880 +2880 @@
-static VALUE
+VALUE
@@ -2904 +2904 @@
-static VALUE
+VALUE
@@ -2974 +2974 @@
-static VALUE
+VALUE
@@ -3008 +3008 @@
-static VALUE
+VALUE
@@ -3044 +3044 @@
-static VALUE
+VALUE
@@ -3129 +3129 @@
-static VALUE
+VALUE
@@ -3164 +3164 @@
-static VALUE
+VALUE
@@ -3245 +3245 @@
-static VALUE
+VALUE
@@ -3275 +3275 @@
-static VALUE
+VALUE
@@ -3407 +3407 @@
-static VALUE
+VALUE
@@ -3434 +3434 @@
-static VALUE
+VALUE
@@ -3540 +3540 @@
-static VALUE
+VALUE
@@ -3561 +3561 @@
-static VALUE
+VALUE
@@ -3590 +3590 @@
-static VALUE
+VALUE
@@ -3650 +3650 @@
-static VALUE
+VALUE
@@ -3721 +3721 @@
-static VALUE
+VALUE
@@ -3864 +3864 @@
-static VALUE
+VALUE
@@ -3999 +3999 @@
-static VALUE
+VALUE
@@ -4043 +4043 @@
-static VALUE
+VALUE
@@ -4206 +4206 @@
-static VALUE
+VALUE
@@ -4242 +4242 @@
-static VALUE
+VALUE
@@ -4264 +4264 @@
-static VALUE
+VALUE
@@ -4468 +4468 @@
-static VALUE
+VALUE
@@ -4516 +4516 @@
-static VALUE
+VALUE
@@ -4569 +4569 @@
-static VALUE
+VALUE
@@ -4650 +4650 @@
-static VALUE
+VALUE
@@ -4687 +4687 @@
-static VALUE
+VALUE
@@ -4739 +4739 @@
-static VALUE
+VALUE
@@ -4792 +4792 @@
-static VALUE
+VALUE
@@ -4857 +4857 @@
-static VALUE
+VALUE
@@ -4909 +4909 @@
-static VALUE
+VALUE
@@ -4942 +4942 @@
-static VALUE
+VALUE
@@ -4975 +4975 @@
-static VALUE
+VALUE
@@ -5004 +5004 @@
-static VALUE
+VALUE
@@ -5112 +5112 @@
-static VALUE
+VALUE
@@ -5157 +5157 @@
-static VALUE
+VALUE
@@ -5196 +5196 @@
-static VALUE
+VALUE
@@ -5247 +5247 @@
-static VALUE
+VALUE
@@ -5283 +5283 @@
-static VALUE
+VALUE
@@ -5425 +5425 @@
-static VALUE
+VALUE
@@ -5461 +5461 @@
-static VALUE
+VALUE
@@ -5595 +5595 @@
-static VALUE
+VALUE
@@ -5632 +5632 @@
-static VALUE
+VALUE
@@ -5691 +5691 @@
-static VALUE
+VALUE
@@ -5725 +5725 @@
-static VALUE
+VALUE
@@ -5792 +5792 @@
-static VALUE
+VALUE
@@ -5830 +5830 @@
-static VALUE
+VALUE
@@ -5886 +5886 @@
-static VALUE
+VALUE
@@ -5924 +5924 @@
-static VALUE
+VALUE
@@ -5980 +5980 @@
-static VALUE
+VALUE
@@ -6083 +6083 @@
-static VALUE
+VALUE
@@ -6110 +6110 @@
-static VALUE
+VALUE
@@ -6138 +6138 @@
-static VALUE
+VALUE
@@ -6170 +6170 @@
-static VALUE
+VALUE
@@ -6190 +6190 @@
-static VALUE
+VALUE
@@ -6226 +6226 @@
-static VALUE
+VALUE
@@ -6262 +6262 @@
-static VALUE
+VALUE
@@ -6298 +6298 @@
-static VALUE
+VALUE
@@ -6352 +6352 @@
-static VALUE
+VALUE
@@ -6424 +6424 @@
-static VALUE
+VALUE
--- string.c
+++ string.c
@@ -67 +67 @@
-static VALUE rb_str_clear(VALUE str);
+VALUE rb_str_clear(VALUE str);
@@ -192 +192 @@
-static VALUE rb_str_downcase(int argc, VALUE *argv, VALUE str);
+VALUE rb_str_downcase(int argc, VALUE *argv, VALUE str);
@@ -1589 +1589 @@
-static VALUE
+VALUE
@@ -1900 +1900 @@
-static VALUE
+VALUE
@@ -1917 +1917 @@
-static VALUE
+VALUE
@@ -2050 +2050 @@
-static VALUE
+VALUE
@@ -2649 +2649 @@
-static VALUE
+VALUE
@@ -2669 +2669 @@
-static VALUE
+VALUE
@@ -3044 +3044 @@
-static VALUE
+VALUE
@@ -3157 +3157 @@
-static VALUE
+VALUE
@@ -3209 +3209 @@
-static VALUE
+VALUE
@@ -3351 +3351 @@
-static VALUE
+VALUE
@@ -3386 +3386 @@
-static VALUE
+VALUE
@@ -3473 +3473 @@
-static VALUE
+VALUE
@@ -3579 +3579 @@
-static VALUE
+VALUE
@@ -3744 +3744 @@
-static VALUE
+VALUE
@@ -3819 +3819 @@
-static VALUE
+VALUE
@@ -3870 +3870 @@
-static VALUE
+VALUE
@@ -3902 +3902 @@
-static VALUE
+VALUE
@@ -4229 +4229 @@
-static VALUE
+VALUE
@@ -4286 +4286 @@
-static VALUE
+VALUE
@@ -4475 +4475 @@
-static VALUE
+VALUE
@@ -4589 +4589 @@
-static VALUE
+VALUE
@@ -4830 +4830 @@
-static VALUE
+VALUE
@@ -4863 +4863 @@
-static VALUE
+VALUE
@@ -4898 +4898 @@
-static VALUE
+VALUE
@@ -5006 +5006 @@
-static VALUE
+VALUE
@@ -5161 +5161 @@
-static VALUE
+VALUE
@@ -5311 +5311 @@
-static VALUE
+VALUE
@@ -5362 +5362 @@
-static VALUE
+VALUE
@@ -5401 +5401 @@
-static VALUE
+VALUE
@@ -5425 +5425 @@
-static VALUE
+VALUE
@@ -5437 +5437 @@
-static VALUE
+VALUE
@@ -5456 +5456 @@
-static VALUE
+VALUE
@@ -5616 +5616 @@
-static VALUE
+VALUE
@@ -5637 +5637 @@
-static VALUE
+VALUE
@@ -5696 +5696 @@
-static VALUE
+VALUE
@@ -5735 +5735 @@
-static VALUE
+VALUE
@@ -5769 +5769 @@
-static VALUE
+VALUE
@@ -5795 +5795 @@
-static VALUE
+VALUE
@@ -5812 +5812 @@
-static VALUE
+VALUE
@@ -6331 +6331 @@
-static VALUE rb_str_is_ascii_only_p(VALUE str);
+VALUE rb_str_is_ascii_only_p(VALUE str);
@@ -6343 +6343 @@
-static VALUE
+VALUE
@@ -6624 +6624 @@
-static VALUE
+VALUE
@@ -6671 +6671 @@
-static VALUE
+VALUE
@@ -6690 +6690 @@
-static VALUE
+VALUE
@@ -6774 +6774 @@
-static VALUE
+VALUE
@@ -6801 +6801 @@
-static VALUE
+VALUE
@@ -6837 +6837 @@
-static VALUE
+VALUE
@@ -6857 +6857 @@
-static VALUE
+VALUE
@@ -6891 +6891 @@
-static VALUE
+VALUE
@@ -6961 +6961 @@
-static VALUE rb_str_delete_bang(int,VALUE*,VALUE);
+VALUE rb_str_delete_bang(int,VALUE*,VALUE);
@@ -7219 +7219 @@
-static VALUE
+VALUE
@@ -7261 +7261 @@
-static VALUE
+VALUE
@@ -7364 +7364 @@
-static VALUE
+VALUE
@@ -7440 +7440 @@
-static VALUE
+VALUE
@@ -7457 +7457 @@
-static VALUE
+VALUE
@@ -7548 +7548 @@
-static VALUE
+VALUE
@@ -7565 +7565 @@
-static VALUE
+VALUE
@@ -7585 +7585 @@
-static VALUE
+VALUE
@@ -7621 +7621 @@
-static VALUE
+VALUE
@@ -7806 +7806 @@
-static VALUE
+VALUE
@@ -8241 +8241 @@
-static VALUE
+VALUE
@@ -8266 +8266 @@
-static VALUE
+VALUE
@@ -8273 +8273 @@
-static VALUE
+VALUE
@@ -8308 +8308 @@
-static VALUE
+VALUE
@@ -8326 +8326 @@
-static VALUE
+VALUE
@@ -8333 +8333 @@
-static VALUE
+VALUE
@@ -8386 +8386 @@
-static VALUE
+VALUE
@@ -8404 +8404 @@
-static VALUE
+VALUE
@@ -8460 +8460 @@
-static VALUE
+VALUE
@@ -8479 +8479 @@
-static VALUE
+VALUE
@@ -8535 +8535 @@
-static VALUE
+VALUE
@@ -8610 +8610 @@
-static VALUE
+VALUE
@@ -8628 +8628 @@
-static VALUE
+VALUE
@@ -8662 +8662 @@
-static VALUE
+VALUE
@@ -8697 +8697 @@
-static VALUE
+VALUE
@@ -8845 +8845 @@
-static VALUE
+VALUE
@@ -8879 +8879 @@
-static VALUE
+VALUE
@@ -8925 +8925 @@
-static VALUE
+VALUE
@@ -8963 +8963 @@
-static VALUE
+VALUE
@@ -9015 +9015 @@
-static VALUE
+VALUE
@@ -9052 +9052 @@
-static VALUE
+VALUE
@@ -9081 +9081 @@
-static VALUE
+VALUE
@@ -9125 +9125 @@
-static VALUE
+VALUE
@@ -9224 +9224 @@
-static VALUE
+VALUE
@@ -9272 +9272 @@
-static VALUE
+VALUE
@@ -9296 +9296 @@
-static VALUE
+VALUE
@@ -9363 +9363 @@
-static VALUE
+VALUE
@@ -9449 +9449 @@
-static VALUE
+VALUE
@@ -9610 +9610 @@
-static VALUE
+VALUE
@@ -9630 +9630 @@
-static VALUE
+VALUE
@@ -9650 +9650 @@
-static VALUE
+VALUE
@@ -9671 +9671 @@
-static VALUE
+VALUE
@@ -9711 +9711 @@
-static VALUE
+VALUE
@@ -9763 +9763 @@
-static VALUE
+VALUE
@@ -9798 +9798 @@
-static VALUE
+VALUE
@@ -9863 +9863 @@
-static VALUE
+VALUE
@@ -9885 +9885 @@
-static VALUE
+VALUE
@@ -9941 +9941 @@
-static VALUE
+VALUE
@@ -9971 +9971 @@
-static VALUE
+VALUE
@@ -10011 +10011 @@
-static VALUE
+VALUE
@@ -10027 +10027 @@
-static VALUE
+VALUE
@@ -10048 +10048 @@
-static VALUE
+VALUE
@@ -10066 +10066 @@
-static VALUE
+VALUE
@@ -10144 +10144 @@
-static VALUE enc_str_scrub(rb_encoding *enc, VALUE str, VALUE repl, int cr);
+VALUE enc_str_scrub(rb_encoding *enc, VALUE str, VALUE repl, int cr);
@@ -10170 +10170 @@
-static VALUE
+VALUE
@@ -10428 +10428 @@
-static VALUE
+VALUE
@@ -10450 +10450 @@
-static VALUE
+VALUE
@@ -10500 +10500 @@
-static VALUE
+VALUE
@@ -10513 +10513 @@
-static VALUE
+VALUE
@@ -10536 +10536 @@
-static VALUE
+VALUE
--- numeric.c
+++ numeric.c
@@ -166 +166 @@
-static VALUE int_even_p(VALUE x);
+VALUE int_even_p(VALUE x);
@@ -170 +170 @@
-static VALUE flo_to_i(VALUE num);
+VALUE flo_to_i(VALUE num);
@@ -397 +397 @@
-static VALUE
+VALUE
@@ -479 +479 @@
-static VALUE
+VALUE
@@ -500 +500 @@
-static VALUE
+VALUE
@@ -516 +516 @@
-static VALUE
+VALUE
@@ -532 +532 @@
-static VALUE
+VALUE
@@ -549 +549 @@
-static VALUE
+VALUE
@@ -562 +562 @@
-static VALUE
+VALUE
@@ -580 +580 @@
-static VALUE
+VALUE
@@ -598 +598 @@
-static VALUE
+VALUE
@@ -616 +616 @@
-static VALUE
+VALUE
@@ -633 +633 @@
-static VALUE
+VALUE
@@ -691 +691 @@
-static VALUE
+VALUE
@@ -704 +704 @@
-static VALUE
+VALUE
@@ -720 +720 @@
-static VALUE
+VALUE
@@ -740 +740 @@
-static VALUE
+VALUE
@@ -756 +756 @@
-static VALUE
+VALUE
@@ -789 +789 @@
-static VALUE
+VALUE
@@ -804 +804 @@
-static VALUE
+VALUE
@@ -817 +817 @@
-static VALUE
+VALUE
@@ -834 +834 @@
-static VALUE
+VALUE
@@ -847 +847 @@
-static VALUE
+VALUE
@@ -870 +870 @@
-static VALUE
+VALUE
@@ -911 +911 @@
-static VALUE
+VALUE
@@ -995 +995 @@
-static VALUE
+VALUE
@@ -1045 +1045 @@
-static VALUE
+VALUE
@@ -1123 +1123 @@
-static VALUE
+VALUE
@@ -1155 +1155 @@
-static VALUE
+VALUE
@@ -1223 +1223 @@
-static VALUE
+VALUE
@@ -1262 +1262 @@
-static VALUE
+VALUE
@@ -1331 +1331 @@
-static VALUE
+VALUE
@@ -1350 +1350 @@
-static VALUE
+VALUE
@@ -1415 +1415 @@
-static VALUE
+VALUE
@@ -1451 +1451 @@
-static VALUE
+VALUE
@@ -1536 +1536 @@
-static VALUE
+VALUE
@@ -1573 +1573 @@
-static VALUE
+VALUE
@@ -1610 +1610 @@
-static VALUE
+VALUE
@@ -1674 +1674 @@
-static VALUE
+VALUE
@@ -1708 +1708 @@
-static VALUE
+VALUE
@@ -1726 +1726 @@
-static VALUE
+VALUE
@@ -1834 +1834 @@
-static VALUE
+VALUE
@@ -1885 +1885 @@
-static VALUE
+VALUE
@@ -1933 +1933 @@
-static VALUE
+VALUE
@@ -2002 +2002 @@
-static VALUE
+VALUE
@@ -2255 +2255 @@
-static VALUE
+VALUE
@@ -2345 +2345 @@
-static VALUE
+VALUE
@@ -2379 +2379 @@
-static VALUE
+VALUE
@@ -2395 +2395 @@
-static VALUE
+VALUE
@@ -2409 +2409 @@
-static VALUE
+VALUE
@@ -2427 +2427 @@
-static VALUE
+VALUE
@@ -2444 +2444 @@
-static VALUE
+VALUE
@@ -2461 +2461 @@
-static VALUE
+VALUE
@@ -2478 +2478 @@
-static VALUE
+VALUE
@@ -2672 +2672 @@
-static VALUE
+VALUE
@@ -2740 +2740 @@
-static VALUE
+VALUE
@@ -3185 +3185 @@
-static VALUE
+VALUE
@@ -3198 +3198 @@
-static VALUE
+VALUE
@@ -3235 +3235 @@
-static VALUE
+VALUE
@@ -3259 +3259 @@
-static VALUE
+VALUE
@@ -3273 +3273 @@
-static VALUE
+VALUE
@@ -3287 +3287 @@
-static VALUE
+VALUE
@@ -3386 +3386 @@
-static VALUE
+VALUE
@@ -3444 +3444 @@
-static VALUE
+VALUE
@@ -3540 +3540 @@
-static VALUE
+VALUE
@@ -3905 +3905 @@
-static VALUE
+VALUE
@@ -4321 +4321 @@
-static VALUE
+VALUE
@@ -4361 +4361 @@
-static VALUE
+VALUE
@@ -4394 +4394 @@
-static VALUE
+VALUE
@@ -4493 +4493 @@
-static VALUE
+VALUE
@@ -4528 +4528 @@
-static VALUE
+VALUE
@@ -4621 +4621 @@
-static VALUE
+VALUE
@@ -4678 +4678 @@
-static VALUE
+VALUE
@@ -4699 +4699 @@
-static VALUE
+VALUE
@@ -4777 +4777 @@
-static VALUE
+VALUE
@@ -4843 +4843 @@
-static VALUE
+VALUE
@@ -4899 +4899 @@
-static VALUE
+VALUE
@@ -4930 +4930 @@
-static VALUE
+VALUE
@@ -4978 +4978 @@
-static VALUE
+VALUE
@@ -4984 +4984 @@
-static VALUE
+VALUE
@@ -5024 +5024 @@
-static VALUE
+VALUE
@@ -5030 +5030 @@
-static VALUE
+VALUE
@@ -5068 +5068 @@
-static VALUE
+VALUE
@@ -5080 +5080 @@
-static VALUE
+VALUE
@@ -5137 +5137 @@
-static VALUE
+VALUE
@@ -5172 +5172 @@
-static VALUE
+VALUE
@@ -5204 +5204 @@
-static VALUE
+VALUE
@@ -5236 +5236 @@
-static VALUE
+VALUE
