diff --git gc.c gc.c
index 346a77ec63..172373e40d 100644
--- gc.c
+++ gc.c
@@ -2935,7 +2935,27 @@ cc_table_mark_i(ID id, VALUE ccs_ptr, void *data_ptr)
         return ID_TABLE_DELETE;
     }
     else {
-        gc_mark(data->objspace, (VALUE)ccs->cme);
+        VALUE cme_obj = (VALUE)ccs->cme;
+
+        if (getenv("STRIPE_DEBUG_T_NONE") && UNLIKELY(RB_TYPE_P(cme_obj, T_NONE))) {
+            VALUE class_path = rb_class_path_cached(data->klass);
+            if (!NIL_P(class_path)) {
+                fprintf(
+                    stdout,
+                    "ERROR (cc_table_mark_i): will fail shortly with try to mark T_NONE object -- obj:%s, P:%s\n",
+                    rb_id2name(id),
+                    RSTRING_PTR(class_path)
+                );
+            } else {
+                fprintf(
+                    stdout,
+                    "ERROR (cc_table_mark_i): will fail shortly fail with try to mark T_NONE object -- obj:%s\n",
+                    rb_id2name(id)
+                );
+            }
+        }
+
+        gc_mark(data->objspace, cme_obj);
 
         for (int i=0; i<ccs->len; i++) {
             VM_ASSERT(data->klass == ccs->entries[i].cc->klass);
