--- ruby-2.4.3-dirty/rbconfig.rb	1969-12-31 16:00:00.000000000 -0800
+++ ruby-2.4.3/rbconfig.rb	2019-05-31 13:50:32.000000000 -0700
@@ -0,0 +1,314 @@
+# encoding: ascii-8bit
+# frozen-string-literal: false
+#
+# The module storing Ruby interpreter configurations on building.
+#
+# This file was created by mkconfig.rb when ruby was built.  It contains
+# build information for ruby which is used e.g. by mkmf to build
+# compatible native extensions.  Any changes made to this file will be
+# lost the next time ruby is built.
+
+module RbConfig
+  RUBY_VERSION.start_with?("2.4.") or
+    raise "ruby lib version (2.4.3) doesn't match executable version (#{RUBY_VERSION})"
+
+  # Ruby installed directory.
+  TOPDIR = File.dirname(__FILE__).chomp!("/lib/ruby/2.4.0/x86_64-darwin18")
+  # DESTDIR on make install.
+  DESTDIR = '' unless defined? DESTDIR
+  # The hash configurations stored.
+  CONFIG = {}
+  CONFIG["DESTDIR"] = DESTDIR
+  CONFIG["MAJOR"] = "2"
+  CONFIG["MINOR"] = "4"
+  CONFIG["TEENY"] = "3"
+  CONFIG["PATCHLEVEL"] = "205"
+  CONFIG["INSTALL"] = '/usr/local/bin/ginstall -c'
+  CONFIG["EXEEXT"] = ""
+  CONFIG["prefix"] = (TOPDIR || DESTDIR + "/Users/trevor/Downloads/ruby-2.4.3/build")
+  CONFIG["ruby_install_name"] = "$(RUBY_BASE_NAME)"
+  CONFIG["RUBY_INSTALL_NAME"] = "$(RUBY_BASE_NAME)"
+  CONFIG["RUBY_SO_NAME"] = "$(RUBY_BASE_NAME)"
+  CONFIG["exec"] = "exec"
+  CONFIG["ruby_pc"] = "ruby-2.4.pc"
+  CONFIG["PACKAGE"] = "ruby"
+  CONFIG["BUILTIN_TRANSSRCS"] = " enc/trans/newline.c"
+  CONFIG["USE_RUBYGEMS"] = "YES"
+  CONFIG["MANTYPE"] = "doc"
+  CONFIG["NROFF"] = "/usr/bin/nroff"
+  CONFIG["vendorarchhdrdir"] = "$(vendorhdrdir)/$(sitearch)"
+  CONFIG["sitearchhdrdir"] = "$(sitehdrdir)/$(sitearch)"
+  CONFIG["rubyarchhdrdir"] = "$(rubyhdrdir)/$(arch)"
+  CONFIG["vendorhdrdir"] = "$(rubyhdrdir)/vendor_ruby"
+  CONFIG["sitehdrdir"] = "$(rubyhdrdir)/site_ruby"
+  CONFIG["rubyhdrdir"] = "$(includedir)/$(RUBY_VERSION_NAME)"
+  CONFIG["RUBY_SEARCH_PATH"] = ""
+  CONFIG["UNIVERSAL_INTS"] = ""
+  CONFIG["UNIVERSAL_ARCHNAMES"] = ""
+  CONFIG["configure_args"] = " '--prefix=/Users/trevor/Downloads/ruby-2.4.3/build'"
+  CONFIG["CONFIGURE"] = "configure"
+  CONFIG["vendorarchdir"] = "$(vendorlibdir)/$(sitearch)"
+  CONFIG["vendorlibdir"] = "$(vendordir)/$(ruby_version)"
+  CONFIG["vendordir"] = "$(rubylibprefix)/vendor_ruby"
+  CONFIG["sitearchdir"] = "$(sitelibdir)/$(sitearch)"
+  CONFIG["sitelibdir"] = "$(sitedir)/$(ruby_version)"
+  CONFIG["sitedir"] = "$(rubylibprefix)/site_ruby"
+  CONFIG["rubyarchdir"] = "$(rubylibdir)/$(arch)"
+  CONFIG["rubylibdir"] = "$(rubylibprefix)/$(ruby_version)"
+  CONFIG["ruby_version"] = "2.4.0"
+  CONFIG["sitearch"] = "$(arch)"
+  CONFIG["arch"] = "x86_64-darwin18"
+  CONFIG["sitearchincludedir"] = "$(includedir)/$(sitearch)"
+  CONFIG["archincludedir"] = "$(includedir)/$(arch)"
+  CONFIG["sitearchlibdir"] = "$(libdir)/$(sitearch)"
+  CONFIG["archlibdir"] = "$(libdir)/$(arch)"
+  CONFIG["libdirname"] = "libdir"
+  CONFIG["RUBY_EXEC_PREFIX"] = "/Users/trevor/Downloads/ruby-2.4.3/build"
+  CONFIG["RUBY_LIB_VERSION"] = ""
+  CONFIG["RUBY_LIB_VERSION_STYLE"] = "3\t/* full */"
+  CONFIG["RI_BASE_NAME"] = "ri"
+  CONFIG["ridir"] = "$(datarootdir)/$(RI_BASE_NAME)"
+  CONFIG["rubysitearchprefix"] = "$(rubylibprefix)/$(sitearch)"
+  CONFIG["rubyarchprefix"] = "$(rubylibprefix)/$(arch)"
+  CONFIG["MAKEFILES"] = "Makefile GNUmakefile"
+  CONFIG["PLATFORM_DIR"] = ""
+  CONFIG["THREAD_MODEL"] = "pthread"
+  CONFIG["SYMBOL_PREFIX"] = "_"
+  CONFIG["EXPORT_PREFIX"] = ""
+  CONFIG["COMMON_HEADERS"] = ""
+  CONFIG["COMMON_MACROS"] = ""
+  CONFIG["COMMON_LIBS"] = ""
+  CONFIG["MAINLIBS"] = ""
+  CONFIG["ENABLE_SHARED"] = "no"
+  CONFIG["DLDLIBS"] = ""
+  CONFIG["SOLIBS"] = ""
+  CONFIG["LIBRUBYARG_SHARED"] = ""
+  CONFIG["LIBRUBYARG_STATIC"] = "-l$(RUBY_SO_NAME)-static -framework CoreFoundation"
+  CONFIG["LIBRUBYARG"] = "$(LIBRUBYARG_STATIC)"
+  CONFIG["LIBRUBY"] = "$(LIBRUBY_A)"
+  CONFIG["LIBRUBY_ALIASES"] = "lib$(RUBY_SO_NAME).so"
+  CONFIG["LIBRUBY_SO"] = "lib$(RUBY_SO_NAME).so.$(RUBY_PROGRAM_VERSION)"
+  CONFIG["LIBRUBY_A"] = "lib$(RUBY_SO_NAME)-static.a"
+  CONFIG["RUBYW_INSTALL_NAME"] = ""
+  CONFIG["rubyw_install_name"] = ""
+  CONFIG["EXTDLDFLAGS"] = ""
+  CONFIG["EXTLDFLAGS"] = ""
+  CONFIG["strict_warnflags"] = ""
+  CONFIG["warnflags"] = "-Wall -Wextra -Wno-unused-parameter -Wno-parentheses -Wno-long-long -Wno-missing-field-initializers -Wno-tautological-compare -Wno-parentheses-equality -Wno-constant-logical-operand -Wno-self-assign -Wunused-variable -Wimplicit-int -Wpointer-arith -Wwrite-strings -Wdeclaration-after-statement -Wshorten-64-to-32 -Wimplicit-function-declaration -Wdivision-by-zero -Wdeprecated-declarations -Wextra-tokens"
+  CONFIG["debugflags"] = "-ggdb3"
+  CONFIG["optflags"] = "-O3 -fno-fast-math"
+  CONFIG["NULLCMD"] = ":"
+  CONFIG["DLNOBJ"] = "dln.o"
+  CONFIG["INSTALL_STATIC_LIBRARY"] = "yes"
+  CONFIG["EXECUTABLE_EXTS"] = ""
+  CONFIG["ARCHFILE"] = ""
+  CONFIG["LIBRUBY_RELATIVE"] = "no"
+  CONFIG["EXTOUT"] = ".ext"
+  CONFIG["PREP"] = "miniruby$(EXEEXT) ruby-runner$(EXEEXT)"
+  CONFIG["CROSS_COMPILING"] = "no"
+  CONFIG["TEST_RUNNABLE"] = "yes"
+  CONFIG["rubylibprefix"] = "$(libdir)/$(RUBY_BASE_NAME)"
+  CONFIG["setup"] = "Setup"
+  CONFIG["ENCSTATIC"] = ""
+  CONFIG["EXTSTATIC"] = ""
+  CONFIG["STRIP"] = "strip -A -n"
+  CONFIG["TRY_LINK"] = ""
+  CONFIG["PRELOADENV"] = "DYLD_INSERT_LIBRARIES"
+  CONFIG["LIBPATHENV"] = "DYLD_LIBRARY_PATH"
+  CONFIG["RPATHFLAG"] = ""
+  CONFIG["LIBPATHFLAG"] = " -L%s"
+  CONFIG["LINK_SO"] = "\n$(POSTLINK)"
+  CONFIG["ASMEXT"] = "S"
+  CONFIG["LIBEXT"] = "a"
+  CONFIG["DLEXT2"] = ""
+  CONFIG["DLEXT"] = "bundle"
+  CONFIG["LDSHAREDXX"] = "$(CXX) -dynamic -bundle"
+  CONFIG["LDSHARED"] = "$(CC) -dynamic -bundle"
+  CONFIG["CCDLFLAGS"] = "-fno-common"
+  CONFIG["STATIC"] = ""
+  CONFIG["ARCH_FLAG"] = ""
+  CONFIG["DLDFLAGS"] = "-Wl,-undefined,dynamic_lookup -Wl,-multiply_defined,suppress"
+  CONFIG["ALLOCA"] = ""
+  CONFIG["codesign"] = "codesign"
+  CONFIG["POSTLINK"] = "test -z '$(RUBY_CODESIGN)' || codesign -s '$(RUBY_CODESIGN)' -f $@"
+  CONFIG["WERRORFLAG"] = "-Werror"
+  CONFIG["CHDIR"] = "cd -P"
+  CONFIG["RMALL"] = "rm -fr"
+  CONFIG["RMDIRS"] = "rmdir -p"
+  CONFIG["RMDIR"] = "rmdir"
+  CONFIG["CP"] = "cp"
+  CONFIG["RM"] = "rm -f"
+  CONFIG["PKG_CONFIG"] = "pkg-config"
+  CONFIG["PYTHON"] = ""
+  CONFIG["DOXYGEN"] = ""
+  CONFIG["DOT"] = "dot"
+  CONFIG["MAKEDIRS"] = "/usr/local/bin/gmkdir -p"
+  CONFIG["MKDIR_P"] = "/usr/local/bin/gmkdir -p"
+  CONFIG["INSTALL_DATA"] = "$(INSTALL) -m 644"
+  CONFIG["INSTALL_SCRIPT"] = "$(INSTALL)"
+  CONFIG["INSTALL_PROGRAM"] = "$(INSTALL)"
+  CONFIG["SET_MAKE"] = ""
+  CONFIG["LN_S"] = "ln -s"
+  CONFIG["NM"] = "nm"
+  CONFIG["DLLWRAP"] = ""
+  CONFIG["WINDRES"] = ""
+  CONFIG["OBJCOPY"] = ":"
+  CONFIG["OBJDUMP"] = "objdump"
+  CONFIG["ASFLAGS"] = ""
+  CONFIG["AS"] = "as"
+  CONFIG["ARFLAGS"] = "rcu "
+  CONFIG["AR"] = "ar"
+  CONFIG["RANLIB"] = "ranlib"
+  CONFIG["try_header"] = ""
+  CONFIG["CC_VERSION_MESSAGE"] = "Apple LLVM version 10.0.1 (clang-1001.0.46.4)\nTarget: x86_64-apple-darwin18.6.0\nThread model: posix\nInstalledDir: /Library/Developer/CommandLineTools/usr/bin"
+  CONFIG["CC_VERSION"] = "$(CC) --version"
+  CONFIG["CSRCFLAG"] = ""
+  CONFIG["COUTFLAG"] = "-o "
+  CONFIG["OUTFLAG"] = "-o "
+  CONFIG["CPPOUTFILE"] = "-o conftest.i"
+  CONFIG["GNU_LD"] = "no"
+  CONFIG["LD"] = "ld"
+  CONFIG["GCC"] = "yes"
+  CONFIG["EGREP"] = "/usr/bin/grep -E"
+  CONFIG["GREP"] = "/usr/bin/grep"
+  CONFIG["CPP"] = "$(CC) -E"
+  CONFIG["CXXFLAGS"] = "$(cxxflags)"
+  CONFIG["OBJEXT"] = "o"
+  CONFIG["CPPFLAGS"] = "-D_XOPEN_SOURCE -D_DARWIN_C_SOURCE -D_DARWIN_UNLIMITED_SELECT -D_REENTRANT $(DEFS) $(cppflags)"
+  CONFIG["LDFLAGS"] = "-L. -fstack-protector -L/usr/local/lib"
+  CONFIG["CFLAGS"] = "$(cflags)  -pipe"
+  CONFIG["CXX"] = "clang++"
+  CONFIG["CC"] = "clang"
+  CONFIG["NACL_LIB_PATH"] = ""
+  CONFIG["NACL_SDK_VARIANT"] = ""
+  CONFIG["NACL_SDK_ROOT"] = ""
+  CONFIG["NACL_TOOLCHAIN"] = ""
+  CONFIG["target_os"] = "darwin18"
+  CONFIG["target_vendor"] = "apple"
+  CONFIG["target_cpu"] = "x86_64"
+  CONFIG["target"] = "x86_64-apple-darwin18"
+  CONFIG["host_os"] = "darwin18.6.0"
+  CONFIG["host_vendor"] = "apple"
+  CONFIG["host_cpu"] = "x86_64"
+  CONFIG["host"] = "x86_64-apple-darwin18.6.0"
+  CONFIG["RUBY_VERSION_NAME"] = "$(RUBY_BASE_NAME)-$(ruby_version)"
+  CONFIG["RUBYW_BASE_NAME"] = "rubyw"
+  CONFIG["RUBY_BASE_NAME"] = "ruby"
+  CONFIG["build_os"] = "darwin18.6.0"
+  CONFIG["build_vendor"] = "apple"
+  CONFIG["build_cpu"] = "x86_64"
+  CONFIG["build"] = "x86_64-apple-darwin18.6.0"
+  CONFIG["RUBY_PROGRAM_VERSION"] = "2.4.3"
+  CONFIG["cxxflags"] = "$(optflags) $(debugflags) $(warnflags)"
+  CONFIG["cppflags"] = ""
+  CONFIG["cflags"] = "$(optflags) $(debugflags) $(warnflags)"
+  CONFIG["target_alias"] = ""
+  CONFIG["host_alias"] = ""
+  CONFIG["build_alias"] = ""
+  CONFIG["LIBS"] = "-lpthread -ldl -lobjc"
+  CONFIG["ECHO_T"] = ""
+  CONFIG["ECHO_N"] = ""
+  CONFIG["ECHO_C"] = "\\\\c"
+  CONFIG["DEFS"] = ""
+  CONFIG["mandir"] = "$(datarootdir)/man"
+  CONFIG["localedir"] = "$(datarootdir)/locale"
+  CONFIG["libdir"] = "$(exec_prefix)/lib"
+  CONFIG["psdir"] = "$(docdir)"
+  CONFIG["pdfdir"] = "$(docdir)"
+  CONFIG["dvidir"] = "$(docdir)"
+  CONFIG["htmldir"] = "$(docdir)"
+  CONFIG["infodir"] = "$(datarootdir)/info"
+  CONFIG["docdir"] = "$(datarootdir)/doc/$(PACKAGE)"
+  CONFIG["oldincludedir"] = "$(SDKROOT)""/usr/include"
+  CONFIG["includedir"] = "$(prefix)/include"
+  CONFIG["localstatedir"] = "$(prefix)/var"
+  CONFIG["sharedstatedir"] = "$(prefix)/com"
+  CONFIG["sysconfdir"] = "$(prefix)/etc"
+  CONFIG["datadir"] = "$(datarootdir)"
+  CONFIG["datarootdir"] = "$(prefix)/share"
+  CONFIG["libexecdir"] = "$(exec_prefix)/libexec"
+  CONFIG["sbindir"] = "$(exec_prefix)/sbin"
+  CONFIG["bindir"] = "$(exec_prefix)/bin"
+  CONFIG["exec_prefix"] = "$(prefix)"
+  CONFIG["PACKAGE_URL"] = ""
+  CONFIG["PACKAGE_BUGREPORT"] = ""
+  CONFIG["PACKAGE_STRING"] = ""
+  CONFIG["PACKAGE_VERSION"] = ""
+  CONFIG["PACKAGE_TARNAME"] = ""
+  CONFIG["PACKAGE_NAME"] = ""
+  CONFIG["PATH_SEPARATOR"] = ":"
+  CONFIG["SHELL"] = "/bin/sh"
+  CONFIG["UNICODE_VERSION"] = "9.0.0"
+  CONFIG["SDKROOT"] = ENV["SDKROOT"] || "" # don't run xcrun everytime, usually useless.
+  CONFIG["archdir"] = "$(rubyarchdir)"
+  CONFIG["topdir"] = File.dirname(__FILE__)
+  # Almost same with CONFIG. MAKEFILE_CONFIG has other variable
+  # reference like below.
+  #
+  #   MAKEFILE_CONFIG["bindir"] = "$(exec_prefix)/bin"
+  #
+  # The values of this constant is used for creating Makefile.
+  #
+  #   require 'rbconfig'
+  #
+  #   print <<-END_OF_MAKEFILE
+  #   prefix = #{Config::MAKEFILE_CONFIG['prefix']}
+  #   exec_prefix = #{Config::MAKEFILE_CONFIG['exec_prefix']}
+  #   bindir = #{Config::MAKEFILE_CONFIG['bindir']}
+  #   END_OF_MAKEFILE
+  #
+  #   => prefix = /usr/local
+  #      exec_prefix = $(prefix)
+  #      bindir = $(exec_prefix)/bin  MAKEFILE_CONFIG = {}
+  #
+  # RbConfig.expand is used for resolving references like above in rbconfig.
+  #
+  #   require 'rbconfig'
+  #   p Config.expand(Config::MAKEFILE_CONFIG["bindir"])
+  #   # => "/usr/local/bin"
+  MAKEFILE_CONFIG = {}
+  CONFIG.each{|k,v| MAKEFILE_CONFIG[k] = v.dup}
+
+  # call-seq:
+  #
+  #   RbConfig.expand(val)         -> string
+  #   RbConfig.expand(val, config) -> string
+  #
+  # expands variable with given +val+ value.
+  #
+  #   RbConfig.expand("$(bindir)") # => /home/foobar/all-ruby/ruby19x/bin
+  def RbConfig::expand(val, config = CONFIG)
+    newval = val.gsub(/\$\$|\$\(([^()]+)\)|\$\{([^{}]+)\}/) {
+      var = $&
+      if !(v = $1 || $2)
+	'$'
+      elsif key = config[v = v[/\A[^:]+(?=(?::(.*?)=(.*))?\z)/]]
+	pat, sub = $1, $2
+	config[v] = false
+	config[v] = RbConfig::expand(key, config)
+	key = key.gsub(/#{Regexp.quote(pat)}(?=\s|\z)/n) {sub} if pat
+	key
+      else
+	var
+      end
+    }
+    val.replace(newval) unless newval == val
+    val
+  end
+  CONFIG.each_value do |val|
+    RbConfig::expand(val)
+  end
+
+  # call-seq:
+  #
+  #   RbConfig.ruby -> path
+  #
+  # returns the absolute pathname of the ruby command.
+  def RbConfig.ruby
+    File.join(
+      RbConfig::CONFIG["bindir"],
+      RbConfig::CONFIG["ruby_install_name"] + RbConfig::CONFIG["EXEEXT"]
+    )
+  end
+end
+CROSS_COMPILING = nil unless defined? CROSS_COMPILING
