diff --git a/thread_pthread.c b/thread_pthread.c
index 7918d0d3d9..a5568885e0 100644
--- a/thread_pthread.c
+++ b/thread_pthread.c
@@ -1530,6 +1530,7 @@ thread_sched_atfork(struct rb_thread_sched *sched)
     }
     vm->ractor.sched.running_cnt = 0;
 
+    rb_native_mutex_initialize(&vm->ractor.sched.lock);
     // rb_native_cond_destroy(&vm->ractor.sched.cond);
     rb_native_cond_initialize(&vm->ractor.sched.cond);
     rb_native_cond_initialize(&vm->ractor.sched.barrier_complete_cond);
